pixy/consumer_test.go

```diff
@@ -0,0 +1,543 @@
+
+	sc1, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+	log.Infof("*** GIVEN 1")
+	// Consume first message to make `consumer-1` subscribe for `test.1`
+	consumed := s.consume(c, sc1, "group-1", "test.1", 2)
+	s.assertMsg(c, consumed[""][0], produced[""][0])
+
+	// When:
+	log.Infof("*** WHEN")
+	sc2, err := SpawnSmartConsumer(s.newConfig("consumer-2"))
+	c.Assert(err, IsNil)
+	_, err = sc2.Consume("group-1", "test.1")
+
+	// Then: `consumer-2` request times out, when `consumer-1` requests keep
+	// return messages.
+	log.Infof("*** THEN")
+	if _, ok := err.(ErrConsumerRequestTimeout); !ok {
+		c.Errorf("Expected ErrConsumerRequestTimeout, got %s", err)
+	}
+	s.consume(c, sc1, "group-1", "test.1", 1, consumed)
+	s.assertMsg(c, consumed[""][1], produced[""][1])
+
+	sc1.Stop()
+	sc2.Stop()
+}
+
+// When a new consumer joins a group the partitions get evenly redistributed
+// among all consumers.
+func (s *SmartConsumerSuite) TestRebalanceOnJoin(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.4")
+	GenMessages(c, "join", "test.4", map[string]int{"A": 10, "B": 10})
+
+	sc1, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+
+	// Consume the first message to make the consumer join the group and
+	// subscribe to the topic.
+	log.Infof("*** GIVEN 1")
+	consumed1 := s.consume(c, sc1, "group-1", "test.4", 1)
+	// Wait until first messages from partitions `A` and `B` are fetched.
+	waitFirstFetched(sc1, 2)
+
+	// Consume 4 messages and make sure that there messages from both
+	// partitions among them.
+	log.Infof("*** GIVEN 2")
+	consumed1 = s.consume(c, sc1, "group-1", "test.4", 4, consumed1)
+	c.Assert(len(consumed1["A"]), Not(Equals), 0)
+	c.Assert(len(consumed1["B"]), Not(Equals), 0)
+	consumedBeforeJoin := len(consumed1["B"])
+
+	// When: another consumer joins the group rebalancing occurs.
+	log.Infof("*** WHEN")
+	sc2, err := SpawnSmartConsumer(s.newConfig("consumer-2"))
+	c.Assert(err, IsNil)
+
+	// Then:
+	log.Infof("*** THEN")
+	consumed2 := s.consume(c, sc2, "group-1", "test.4", consumeAll)
+	consumed1 = s.consume(c, sc1, "group-1", "test.4", consumeAll, consumed1)
+	// Partition "A" has been consumed by `consumer-1` only
+	c.Assert(len(consumed1["A"]), Equals, 10)
+	c.Assert(len(consumed2["A"]), Equals, 0)
+	// Partition "B" has been consumed by both consumers, but ever since
+	// `consumer-2` joined the group the first one have not got any new messages.
+	c.Assert(len(consumed1["B"]), Equals, consumedBeforeJoin)
+	c.Assert(len(consumed2["B"]), Not(Equals), 0)
+	c.Assert(len(consumed1["B"])+len(consumed2["B"]), Equals, 10)
+	// `consumer-2` started consumer from where `consumer-1` left off.
+	c.Assert(consumed2["B"][0].Offset, Equals, consumed1["B"][len(consumed1["B"])-1].Offset+1)
+
+	sc2.Stop()
+	sc1.Stop()
+}
+
+// When a consumer leaves a group the partitions get evenly redistributed
+// among the remaining consumers.
+func (s *SmartConsumerSuite) TestRebalanceOnLeave(c *C) {
```

Comment:
This test failed on Travis CI on the most recent commit on Go 1.4, but passed on Go 1.5. Looks like it might be flaky.